src/main/perf/IndexThreads.java

```diff
@@ -241,10 +241,24 @@ public void run() {
           }
           final double docsPerGroupBlock = numTotalDocs / (double) groupBlocks.length;
 
-          while (stop.get() == false && docs.reserve()) {
-            final int groupCounter = groupBlockIndex.getAndIncrement();
-            if (groupCounter >= groupBlocks.length) {
+          while (stop.get() == false) {
+            int groupCounter = -1;
+            if (groupBlockIndex.get() >= groupBlocks.length) {
+              docs.recycle();
               break;
+            } else {
+              synchronized (groupBlockIndex) {
+                // we need to make sure we have more group index
+                // as well as more docs to index at the same time
+                if (groupBlockIndex.get() >= groupBlocks.length) {
+                  docs.recycle();
+                  break;
+                }
+                if (docs.reserve() == false) {
+                  break;
+                }
+                groupCounter = groupBlockIndex.getAndIncrement();
```

Comment:
Since we are sync'd anyway, we could probably make this an ordinary `int groupBlockIndex` now, but let's defer that.

Reply:
It's actually trickier than I thought, I tried to use `Integer` instead once but then found out it is value-based so basically not able to be shared and edited among multiple threads. 
Another option is to make it a member of `IndexThreads` and pass that parent object in, this should work but I'm too lazy so just keep it the old way LOL.

Reply:
LOL ok good!